### PR TITLE
Test Coverage for FDI version verification

### PR DESCRIPTION
**Pull request by adarshdubey-star**
### Problem Statement
The versioning of the foreman discovery image should be consistent or greater in the latest version of the satellite.

### Solution
Save the FDI version pre_upgrade and post upgrading check the version.

### Related Issues
[Bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=2239115)

### CODE REVIEW

File: tests/upgrades/test_discovery.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import pytest
+
+
+class TestDiscoveryImage:
+    """Pre-upgrade and post-upgrade scenarios to test Foreman Discovery Image version.
+
+    Test Steps:
+        1. Before Satellite upgrade, Check the FDI version on the Discovered host
+        2. Upgrade satellite.
+        3. Check the FDI version on the Discovered host, after upgrade to make sure its same or greater.
```
Comment (jameerpathan111) [inline tests/upgrades/test_discovery.py:31]:
```suggestion
        3. Check the FDI version on the Discovered host after upgrade to make sure its same or greater.
```

File: tests/upgrades/test_discovery.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            1. Check installed version of FDI
+
+        :expectedresults: Version should be saved and checked post-upgrade
+        """
+        target_sat.register_to_cdn()
+        target_sat.execute(
+            'yum -y --disableplugin=foreman-protector install foreman-discovery-image'
+        )
+        target_sat.unregister()
+        fdi_version = target_sat.execute('rpm -qa *foreman-discovery-image*')
```
Comment (jameerpathan111) [inline tests/upgrades/test_discovery.py:50]: See if `target_sat.execute(f'rpm -q --qf "%{{VERSION}}" {rpm_name}').stdout` works.

**Review by jameerpathan111 — CHANGES_REQUESTED** [review]

File: tests/upgrades/test_discovery.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+        :steps:
+            1. Check installed version of FDI
+
+        :expectedresults: Version should be greater than or equal to pre_upgrade version
+        """
+        pre_upgrade_version = pre_upgrade_data.get('pre_upgrade_version')
+        target_sat.register_to_cdn()
+        target_sat.execute(
+            'yum -y --disableplugin=foreman-protector install foreman-discovery-image'
```
Comment (Gauravtalreja1) [inline tests/upgrades/test_discovery.py:71]: It should be preinstalled from pre-upgrade, and will be updated to equal or greater version during upgrades, right?
why removing it earlier?

File: tests/upgrades/test_discovery.py (hunk 2 above)
Comment (lhellebr) [reply tests/upgrades/test_discovery.py:50]: Firstly, I think it should be "%{VERSION}". Secondly, I think an f-string will have hard time with the curly braces :)

File: tests/upgrades/test_discovery.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        :id: preupgrade-8c94841c-6791-4af0-aa9c-e54c8d8b9a92
+
+        :steps:
+            1. Check installed version of FDI
+
+        :expectedresults: Version should be saved and checked post-upgrade
+        """
+        target_sat.register_to_cdn()
+        target_sat.execute(
+            'yum -y --disableplugin=foreman-protector install foreman-discovery-image'
```
Comment (lhellebr) [inline tests/upgrades/test_discovery.py:47]: Why not do it the documented way, i.e. foreman-maintain?

File: tests/upgrades/test_discovery.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        )
+        target_sat.unregister()
+        fdi_version = target_sat.execute('rpm -qa *foreman-discovery-image*')
+        # Note: The regular exp takes care of format digit.digit.digit or digit.digit.digit-digit in the output
+        pre_upgrade_version = re.search(r'\d+\.\d+\.\d+(-\d+)?', fdi_version.stdout).group()
+        pre_upgrade_version = version.parse(pre_upgrade_version)
+        save_test_data({'pre_upgrade_version': pre_upgrade_version})
+        target_sat.execute('yum -y remove foreman-discovery-image')
```
Comment (lhellebr) [inline tests/upgrades/test_discovery.py:55]: Why do you remove the package? I doubt it's something that should be done pre-upgrade, as per docs.

File: tests/upgrades/test_discovery.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        :steps:
+            1. Check installed version of FDI
+
+        :expectedresults: Version should be saved and checked post-upgrade
+        """
+        target_sat.register_to_cdn()
+        target_sat.execute(
+            'yum -y --disableplugin=foreman-protector install foreman-discovery-image'
+        )
+        target_sat.unregister()
```
Comment (lhellebr) [inline tests/upgrades/test_discovery.py:49]: Shouldn't the registration survive upgrade?

File: tests/upgrades/test_discovery.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        fdi_version = target_sat.execute('rpm -qa *foreman-discovery-image*')
+        # Note: The regular exp takes care of format digit.digit.digit or digit.digit.digit-digit in the output
+        post_upgrade_version = re.search(r'\d+\.\d+\.\d+(-\d+)?', fdi_version.stdout).group()
+        post_upgrade_version = version.parse(post_upgrade_version)
+        assert post_upgrade_version >= pre_upgrade_version
+        target_sat.execute('yum -y remove foreman-discovery-image')
```
Comment (lhellebr) [inline tests/upgrades/test_discovery.py:79]: Why?

File: tests/upgrades/test_discovery.py (hunk 2 above)
Comment (lhellebr) [reply tests/upgrades/test_discovery.py:50]: But yes, try something along the lines of what Jameer suggested. The way you're doing it looks too complicated.

**Review by lhellebr — COMMENTED** [review]
I'm no expert in FDI but a lot of things seem fishy to me, I left quite a few comments.

File: tests/upgrades/test_discovery.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+from packaging import version
+import pytest
+
+
+class TestDiscoveryImage:
+    """Pre-upgrade and post-upgrade scenarios to test Foreman Discovery Image version.
+
+    Test Steps:
+        1. Before Satellite upgrade, Check the FDI version on the Discovered host
```
[the remark this answers is not in the document]
Comment (adarshdubey-star) [reply tests/upgrades/test_discovery.py:29]: We are only concerned about the FDI version installed on the satellite , there are other scenarios which have assertions related to the discovered hosts.

File: tests/upgrades/test_discovery.py (hunk 4 above)
Comment (adarshdubey-star) [reply tests/upgrades/test_discovery.py:47]: This is something that I got inspired from - https://github.com/SatelliteQE/robottelo/blob/35dca5a00df0e6a42a14e00cf46ff5fcc91c7b8b/pytest_fixtures/core/broker.py#L62C26-L62C26

File: tests/upgrades/test_discovery.py (hunk 2 above)
Comment (adarshdubey-star) [reply tests/upgrades/test_discovery.py:50]: Thanks for the input I will try it out.

File: tests/upgrades/test_discovery.py (hunk 5 above)
Comment (adarshdubey-star) [reply tests/upgrades/test_discovery.py:55]: Yeah, I will have a look.

File: tests/upgrades/test_discovery.py (hunk 3 above)
Comment (adarshdubey-star) [reply tests/upgrades/test_discovery.py:71]: Good point.

File: tests/upgrades/test_discovery.py (hunk 7 above)
Comment (adarshdubey-star) [reply tests/upgrades/test_discovery.py:79]: I believe you are talking about the removal of the package post-upgrade, I will check if its really needed.

File: tests/upgrades/test_discovery.py (hunk 2 above)
Comment (adarshdubey-star) [reply tests/upgrades/test_discovery.py:50]: I tried out with the release  as well what I am looking for is the whole version '4.1.0-10' 
```
result = target_sat.execute(f'rpm -q --qf "%{{VERSION}}-%{{RELEASE}}" {"foreman-discovery-image"}').stdout
ipdb> result
'4.1.0-10.el8sat'

```
And again I have split and do some modifications here to extract that. So keeping it as it is.

File: tests/upgrades/test_discovery.py (hunk 7 above)
Comment (adarshdubey-star) [reply tests/upgrades/test_discovery.py]: updated

File: tests/upgrades/test_discovery.py (hunk 4 above)
Comment (adarshdubey-star) [reply tests/upgrades/test_discovery.py]: Yes @lhellebr I have updated it, Thanks for your input.

File: tests/upgrades/test_discovery.py (hunk 6 above)
Comment (adarshdubey-star) [reply tests/upgrades/test_discovery.py]: updated

**Review by lhellebr — APPROVED** [review]

**Pull request merged**